### PR TITLE
fix(matching): use strict object structure matching for object comparisons

### PR DESCRIPTION
**Pull request by jujaga**
# Description

This pull request introduces a new utility function for strict shallow object comparison and refactors parts of the `record` service to use this new function for improved object matching logic. The changes also include comprehensive unit tests for the new function and update type usage for clarity.

**Key changes:**

### New Utility Function

* Added a new `shallowEqual` function to `src/utils/utils.ts` that performs strict, top-level equality checks between two objects, with optional key allowlisting and special handling for date-string equivalence.

### Service Logic Refactoring

* Refactored `replaceRecordService` in `src/services/record.ts` to use `shallowEqual` instead of `containsSubset` when matching old and new event objects, ensuring stricter and more predictable matching logic for coding and process events.
* Updated imports and function signatures in `src/services/record.ts` to use the more explicit `PiesRecord` type instead of the generic `Record` type, improving type safety and clarity.

### Testing

* Added a comprehensive test suite for `shallowEqual` in `tests/unit/utils/utils.test.ts`, covering edge cases such as key mismatches, array and object references, date-string equivalence, and key allowlisting.
* Updated test imports to include `shallowEqual`.

## Related Issues/Tickets

[PSB-176](https://apps.nrs.gov.bc.ca/int/jira/browse/PSB-176)

## Type of Change

- [x] Bug fix (non-breaking change)
- [ ] New feature (non-breaking change)
- [ ] Breaking change (fix or feature that changes existing behavior)
- [ ] Documentation update

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) guidelines
- [x] I have tested my changes locally
- [x] The code builds and passes all tests
- [x] I have updated documentation as needed

## Instance Deployment

- [ ] I require a cloud deployment for testing

> [!NOTE]
> To request for a cloud deployment for testing purposes, add the `deploy` label to this PR.
> Deployments will be removed when the PR is closed, merged, or the `deploy` label is removed.

### CODE REVIEW

File: src/utils/utils.ts [hunk 1]
```diff
@@ -122,6 +122,55 @@ export function getUUIDv7Timestamp(uuid: string): number | undefined {
...
+  const isDateEquivalent = (left: unknown, right: unknown): boolean => {
+    if (typeof left !== 'string' && typeof left !== 'number') return false;
+    if (typeof right !== 'string' && typeof right !== 'number') return false;
+
+    const leftTime = Date.parse(left.toString());
+    const rightTime = Date.parse(right.toString());
+
+    if (Number.isNaN(leftTime) || Number.isNaN(rightTime)) return false;
+    return leftTime === rightTime;
+  };
```
[the remark this answers is not in the document]
Comment (jujaga) [reply src/utils/utils.ts:156]: Dash check deliberately dropped.

File: src/services/record.ts [hunk 2]
```diff
@@ -240,9 +256,26 @@ export const replaceRecordService = (data: Record, principal?: string): Promise<
...
         };
-        const matched = peOld.find((old) => containsSubset(old, peNew));
+        const peMatched = peOld.find((pe) => {
+          const peOld: Record<string, unknown> = {};
+          for (const [key, value] of Object.entries(pe) as [string, unknown][]) {
+            if (key === 'id' || value === undefined) continue;
+            peOld[key] = value;
+          }
+
+          return shallowEqual(peOld, peNew, [
```
[the remark this answers is not in the document]
Comment (jujaga) [reply src/services/record.ts:266]: Intentional behaviour; if any status changes (value change or drop), this must trigger a forced update for the row (via recreation).

File: src/utils/utils.ts [hunk 3]
```diff
@@ -122,6 +122,55 @@ export function getUUIDv7Timestamp(uuid: string): number | undefined {
...
+ * - Treats `null` and `undefined` as equivalent (nullish equivalence).
+ * - Supports an optional allowlist of expected keys.
+ * - When expected keys are provided, any key in `lhs` or `rhs` not in that list fails the comparison.
+ * - Expected keys may be omitted from both objects without failing the comparison.
+ * @param lhs - Left-hand object to compare.
+ * @param rhs - Right-hand object to compare.
+ * @param expectedKeys - Optional allowlist of top-level keys.
+ * @returns `true` when objects are strictly equivalent at the top level and satisfy key constraints.
+ */
+export function shallowEqual(
```
Comment (qhanson55) [inline src/utils/utils.ts:139]: Should consider if we want this to be a generic shallow equating function or more domain specific, something like `compareRecordEvents` or something. Some of the functionality is kinda toeing that line.

Also not sure if you want to keep `containsSubset` still? It's no longer being used anywhere.

File: src/utils/utils.ts [hunk 4]
```diff
@@ -122,6 +122,55 @@ export function getUUIDv7Timestamp(uuid: string): number | undefined {
...
+  lhs: Record<string, unknown>,
+  rhs: Record<string, unknown>,
+  expectedKeys?: string[]
+): boolean {
+  const isDateEquivalent = (left: unknown, right: unknown): boolean => {
+    if (typeof left !== 'string' && typeof left !== 'number') return false;
```
Comment (qhanson55) [inline src/utils/utils.ts:146]: I think we're at risk of false positives here. Something that isn't a date could be equated as equal once we use `Date.parse()` with `.toString()`. Since we know that our dates get converted to strings (`eventToDateParts`) then maybe we can just check for both to be string here? If this doesn't need to be a [generic util function](https://github.com/bcgov/nr-peach/pull/258/changes#r3319482300) then could probably leave as is.

File: src/utils/utils.ts [hunk 5]
```diff
@@ -122,6 +122,55 @@ export function getUUIDv7Timestamp(uuid: string): number | undefined {
...
+
+  const compareValue = (left: unknown, right: unknown): boolean => {
+    if (Object.is(left ?? null, right ?? null)) return true;
+    return isDateEquivalent(left, right);
+  };
+
+  const lhsKeys = Object.keys(lhs);
+  const rhsKeys = Object.keys(rhs);
+
+  if (lhsKeys.length !== rhsKeys.length) return false;
```
Comment (qhanson55) [inline src/utils/utils.ts:165]: This line contradicts the doc line "`Treats `null` and `undefined` as equivalent (nullish equivalence).`". It works cause you're stripping the undefined keys in `replaceRecord` with 
```ts
if (key === 'id' || value === undefined) continue;
```
but if used elsewhere without the knowledge of needing to strip undefined then this could cause issues.

File: src/utils/utils.ts [hunk 6]
```diff
@@ -122,6 +122,55 @@ export function getUUIDv7Timestamp(uuid: string): number | undefined {
...
+
+  if (expectedKeys) {
+    const allowed = new Set(expectedKeys);
```
Comment (qhanson55) [inline src/utils/utils.ts]: Consider keeping as an array: 
If `expectedKeys` is going to be small in size, just leave as an array and use `.includes()`, look up speed of the two with a small array is negligible. And, if `shallowEqual` is used in a nested loop of size `n*m`, the benefit of the set is negated by the memory usage of instantiating `n*m` sets. Even though, in the use case of `replaceRecord` it doesn't make much of a difference either way as event sets are small (`n*m` is small), if this is to be a [generic util function for shallow equating](https://github.com/bcgov/nr-peach/pull/258/changes#r3319482300) then we may want to use an array in case it's used on large loops.

File: tests/unit/utils/utils.test.ts [hunk 7]
```diff
@@ -241,6 +241,79 @@ describe('getUUIDv7Timestamp', () => {
   });
 });
 
+describe('shallowEqual', () => {
```
Comment (qhanson55) [inline tests/unit/utils/utils.test.ts:244]: A test for an edge case test for empty objects and test cases that should be considered based on these comments (depending on generic vs more domain specific):

1. [false positives](https://github.com/bcgov/nr-peach/pull/258/changes#r3319545752) - what should be returned if:
```ts
const lhs = { a: 1 };
const rhs = { a: '1' };
```

2. [null === undefined](https://github.com/bcgov/nr-peach/pull/258/changes#r3319562723) - what result should we expect for this case?
```ts
const lhs = { a: 1, b: undefined };
const rhs = { a: 1 };
```
```ts
expect(shallowEqual(lhs, rhs)).toBe(true);
```
or
```ts
expect(shallowEqual(lhs, rhs)).toBe(false);
```

Comment (jujaga) [reply tests/unit/utils/utils.test.ts:244]: Very good catch - agreed this should be false

File: src/utils/utils.ts (hunk 3 above)
Comment (jujaga) [reply src/utils/utils.ts:139]: Keeping containsSubset for now until we look at merge logic; may provide nominal value in that future flow.

File: src/utils/utils.ts (hunk 6 above)
Comment (jujaga) [reply src/utils/utils.ts]: Reverted back to array as this should typically be a finite set when defined

File: src/utils/utils.ts (hunk 5 above)
Comment (jujaga) [reply src/utils/utils.ts:165]: Object.is does this check; added test and modified jsdocs to clarify intent

**Review by jujaga — COMMENTED** [review]
Thanks for the review - caught a couple extra edge cases that slipped under my radar.

**Review by qhanson55 — APPROVED** [review]
